### PR TITLE
feat: full PayRef implementation

### PR DESCRIPTION
**Pull request by fluffypony**
## Description
---

Implements the Payment Reference (PayRef) system for Tari, providing globally unique identifiers for individual transaction outputs to enable payment verification in Mimblewimble-based blockchains. PayRefs solve the critical business problem of payment verification for exchanges and merchants without compromising privacy.

**Key Features:**
- Generate PayRefs using `PayRef = Blake2b_256(block_hash || output_hash)` for merkle proof compatibility
- Transaction-based PayRef calculation ensuring sender and receiver generate identical references
- Support for multiple PayRefs per transaction (multi-output payments)
- Confirmation-based stability with reorg safety (default: 5 confirmations)
- Privacy-preserving payment verification
- Full FFI support for external integrations
- Exchange documentation and integration examples

## Motivation and Context
---

The primary use case is when users send payments to exchanges but the exchange claims they haven't received the payment. Currently, there's no reliable way to prove payments in Mimblewimble due to:

- **Kernel obfuscation**: No direct link between kernels and specific outputs
- **Receiver limitations**: Receivers never see kernels in one-sided payments  
- **Privacy by design**: Transaction relationships are intentionally hidden

This creates significant business friction:
- High customer support burden for exchanges
- Users cannot prove payments were sent
- Difficulty maintaining audit trails
- Adoption friction due to payment verification problems

PayRefs provide a practical solution that ensures both sender and receiver can generate the same payment reference for verification, while maintaining privacy and enabling future merkle proof verification against block headers.

## How Has This Been Tested?
---

- **Unit tests**: PayRef generation with new formula, multiple output scenarios, change output identification
- **Integration tests**: End-to-end payment flows ensuring sender-receiver PayRef matching
- **Reorg tests**: Verification of PayRef cleanup during blockchain reorganizations
- **Manual testing**: Console wallet transaction history display with multiple PayRefs support
- **API testing**: gRPC endpoints for PayRef retrieval supporting multiple references per transaction
- **Edge case validation**: Handling of spent outputs, burned outputs, and same-block spends

## What process can a PR reviewer use to test or verify this change?
---

**Basic PayRef Generation:**
1. Send a transaction with multiple outputs using the console wallet
2. Wait for 5+ confirmations 
3. Check transaction history - PayRefs should appear as "Available"
4. Verify each PayRef is a 64-character hex string
5. Confirm receiver generates identical PayRefs for the same outputs

**Multi-Output Transactions:**
1. Create a transaction with multiple recipients
2. Verify multiple PayRefs are displayed (one per non-change output)
3. Confirm each recipient can verify their specific PayRef

**PayRef Search:**
1. In console wallet, press `t` for transactions, then `s` to search
2. Enter a known PayRef (full or partial)
3. Verify it finds the matching transaction

**Reorg Safety Testing:**
1. Force a reorg on a test network
2. Verify PayRefs are properly cleaned up for reorged transactions
3. Confirm no orphaned PayRef entries remain

**FFI Interface:**
1. Build with `cargo build`
2. Check that new FFI functions compile without errors
3. Verify header file `wallet.h` includes new PayRef structures with array support

**Documentation:**
1. Review exchange integration guide in `docs/src/09_adding_tari_to_your_exchange.md`
2. Check user guide in `docs/src/payref_userguide.md`
3. Verify code examples reflect the new multi-PayRef support

**API Testing:**
```bash
# Test PayRef retrieval for specific transaction
grpcurl -plaintext -d '{"tx_id": "YOUR_TX_ID"}' \
  localhost:18143 tari.rpc.Wallet/GetTransactionPayRefs

# Test payment lookup by PayRef  
grpcurl -plaintext -d '{"payment_reference_hex": "YOUR_PAYREF_HERE"}' \
  localhost:18143 tari.rpc.Wallet/GetPaymentByReference

# Verify multiple PayRefs for multi-output transactions
grpcurl -plaintext localhost:18143 tari.rpc.Wallet/GetCompletedTransactions
```

## Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork  
- [ ] Other - Please specify

## Summary by CodeRabbit

- **New Features**
  - Introduced Payment Reference (PayRef) support across Tari blockchain outputs, enabling unique, privacy-preserving identifiers.
  - Added CLI commands and UI enhancements in the console wallet for displaying, searching, and verifying PayRefs.
  - Enabled gRPC and FFI APIs for retrieving and verifying transactions by PayRef, facilitating integration with external services.
  - Implemented database schema migrations, indexing, and backend support for efficient PayRef lookups.
  - Added a BaseNode gRPC streaming method to search blockchain outputs by payment references.
  - Extended wallet and transaction services with PayRef generation, storage, and querying capabilities.
  - Added PayRef verification and display configuration options in the wallet UI.

- **Documentation**
  - Added comprehensive user guides and integration documentation for PayRef usage and best practices.
  - Updated application and API overviews to include PayRef features.
  - Added detailed gRPC and FFI overviews covering PayRef functionality.

- **Bug Fixes / Refactor**
  - Improved error handling and code clarity across multiple modules.
  - Streamlined output hash tracking and transaction model updates to support PayRef features.
  - Simplified error constructions and iterator usage in core and communication modules for cleaner code.

- **Chores**
  - Upgraded Diesel dependency in multiple crates for improved compatibility and security.
  - Added new hash domain for Payment Reference generation.
  - Minor code cleanups and simplifications across core and comms crates.

### CODE REVIEW

**Review by stringhandler — DISMISSED** [review]
utack

File: applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs [hunk 1]
```diff
@@ -202,6 +213,110 @@ impl WalletGrpcServer {
...
+/// Private helper function that contains the core payment reference calculation logic
+async fn calculate_payment_references_impl(
+    mut output_manager: OutputManagerHandle,
+    output_commitments: &[Vec<u8>],
+    mined_height: u64,
+) -> Vec<Vec<u8>> {
+    let mut payment_refs = Vec::new();
+    
+    // Note: Don't skip based on transaction mined_height for restored wallets
+    // Individual outputs may have mining data even if transaction doesn't
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs]: This should not be the case, there was a bug that was recently fix where this did occur. The mined height in the transaction and mined_height of the output should be in sync

File: applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs [hunk 2]
```diff
@@ -202,6 +213,110 @@ impl WalletGrpcServer {
...
+        }
+    };
+    
+    let spent_outputs = match output_manager.get_spent_outputs().await {
+        Ok(outputs) => outputs,
+        Err(e) => {
+            warn!(target: LOG_TARGET, "Failed to get spent outputs for PayRef calculation: {}", e);
+            Vec::new()
+        }
+    };
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs]: This will not be scalable, the internal DB in the OMS has methods to ask for specific outputs, we just need to expose those requests. We don't need to pull out the entire DB to memory here.

File: applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs [hunk 3]
```diff
@@ -1103,6 +1223,22 @@ impl wallet_server::Wallet for WalletGrpcServer {
...
+                    txn.mined_height.unwrap_or(0),
+                ).await;
+                
+                let payment_reference = payment_refs
+                    .first()
+                    .cloned()
+                    .unwrap_or_default()
+                    .iter()
+                    .map(|b| format!("{:02x}", b))
+                    .collect::<String>();
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs]: this is wrong in two ways:
1- if there is no payref we should make it none, and not the default. 
2- We should not return the first payref,most transactions have 2 outputs, 1 for change, and 1 for sent. We need to include both or only send the sent output one. 

I would only do the sent out, and do figure out which one is the sent one, we need to search the oms for the one that does not exist.

File: applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs [hunk 4]
```diff
@@ -1200,55 +1337,57 @@ impl wallet_server::Wallet for WalletGrpcServer {
...
+                    Ok(c) => c.as_bytes().to_vec(),
+                    Err(e) => {
+                        warn!(target: LOG_TARGET, "Failed to get input commitment: {}", e);
+                        vec![]
+                    },
+                })
+                .collect();
+            
+            // Calculate PayRef for this transaction's first output (if mined)
+            let payment_refs = self.calculate_payment_references_for_transaction(&output_commitments, txn.mined_height.unwrap_or(0)).await;
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs]: See above, we cant use first output. 
Also if the txn.mined_height is none, or = 0, we can assume its not mined.

File: applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs [hunk 5]
```diff
@@ -1283,57 +1422,61 @@ impl wallet_server::Wallet for WalletGrpcServer {
...
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs]: same as above

File: applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs [hunk 6]
```diff
@@ -1620,7 +1763,190 @@ impl wallet_server::Wallet for WalletGrpcServer {
...
+            ));
+        }
+
+        let payment_ref = message.payment_reference.try_into().map_err(|_| {
+            Status::invalid_argument("payment_reference must be exactly 32 bytes".to_string())
+        })?;
+
+        let mut output_service = self.get_output_manager_service();
+        
+        match output_service.find_payment_by_reference(payment_ref).await {
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs]: This will be very limited, you will only be able to search your own received outputs on payment ref, not sent. 
We need to implement this on the TMS and per sent transaction rather,

File: applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs [hunk 7]
```diff
@@ -202,6 +213,110 @@ impl WalletGrpcServer {
...
+    /// Calculate PayRefs for a transaction's outputs using existing wallet output data
+    async fn calculate_payment_references_for_transaction(
+        &self,
+        output_commitments: &[Vec<u8>],
+        mined_height: u64,
+    ) -> Vec<Vec<u8>> {
+        calculate_payment_references_impl(
+            self.get_output_manager_service(),
+            output_commitments,
+            mined_height,
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/src/grpc/wallet_grpc_server.rs]:
```suggestion
    async fn calculate_payment_references_for_transaction(
        output_manager: &OutputManagerHandle,
        output_commitments: &[Vec<u8>],
        mined_height: u64,
    ) -> Vec<Vec<u8>> {
        calculate_payment_references_impl(
            output_manager,
            output_commitments,
            mined_height,
```
By doing this, you don't have to pass around the grpc wallet service

File: applications/minotari_console_wallet/src/ui/state/app_state.rs [hunk 8]
```diff
@@ -777,10 +777,99 @@ impl AppStateInner {
...
             })
             .collect::<Result<Vec<_>, _>>()?;
+        // Calculate payment references for completed transactions using canonical approach
+        self.calculate_payment_references_for_transactions().await?;
+        
         self.updated = true;
         Ok(())
     }
 
+    async fn calculate_payment_references_for_transactions(&mut self) -> Result<(), UiError> {
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/src/ui/state/app_state.rs]: This is def going to be displaying different results in different wallets.

**Review by SWvheerden — CHANGES_REQUESTED** [review]
I believe there are some fundamental problems in the way this is implemented. 
There are some improvements that need to be made for this to be scalable such as don't query the entire db if you want a single commitment etc. 

But the issue here is in how transactions works. 
When a transaction is created the TMS saves this transaction, the OMS will only save your outputs. So not the sent outputs. So a sender can never create a payref from the OMS as the sent outputs are not stored in there. 

The other issue here is that a tx has on avg 2 outputs, maybe less maybe more. One is a change coming back to you if there is a change, other is for each output going out. 

What we need to do is create a payref from the TMS when we create the tx, it might need to be stored in an incomplete state, so that you can calculate the payref when you want for the tx. 
But the sender when creating the tx will know which tx is its own change, and which is the outputs going out. We need to flag those outputs so we can create payrefs from them. 
We can then just ask the TMS for the payref for a tx, by looking at the sent outputs, and taking the mined_height and calculating the payrefs.

File: base_layer/core/src/base_node/comms_interface/comms_request.rs [hunk 9]
```diff
@@ -83,6 +83,8 @@ pub enum NodeCommsRequest {
     FetchUnspentUtxosInBlock {
         block_hash: BlockHash,
     },
+    FetchOutputByPayRef([u8; 32]),
```
Comment (SWvheerden) [inline base_layer/core/src/base_node/comms_interface/comms_request.rs]:
```suggestion
    FetchOutputByPayRef(FixedHash),
```
we use 32byte hashes all over and parse this as fixedhash so that we don't have to handle conversions everywhere

File: base_layer/core/src/chain_storage/async_db.rs [hunk 10]
```diff
@@ -433,3 +435,23 @@ mod tests {
...
+        }).await.map_err(|_| ChainStorageError::AccessError("Task join error".to_string()))?
+    }
+
+    /// Check if an output is spent and return spent information
+    pub async fn check_output_spent_status(&self, output_hash: HashOutput) -> Result<Option<InputMinedInfo>, ChainStorageError> {
+        let db = self.db.clone();
+        tokio::task::spawn_blocking(move || {
+            let backend = db.db_read_access()?;
+            backend.check_output_spent_status(output_hash)
+        }).await.map_err(|_| ChainStorageError::AccessError("Task join error".to_string()))?
```
Comment (SWvheerden) [inline base_layer/core/src/chain_storage/async_db.rs]: why not use the above macro? `make_async_fn!`

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 11]
```diff
@@ -56,6 +59,7 @@ use tari_utilities::{
     hex::{to_hex, Hex},
     ByteArray,
 };
+// PayRef generation moved to shared utility crate
```
Comment (SWvheerden) [inline base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs]:
```suggestion
```

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 12]
```diff
@@ -30,6 +30,8 @@ use std::{
     time::Instant,
 };
 
+// Blake2b and digest removed - using shared PayRef utility
```
Comment (SWvheerden) [inline base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs]:
```suggestion
```

File: applications/minotari_console_wallet/src/ui/state/app_state.rs [hunk 13]
```diff
@@ -777,10 +777,99 @@ impl AppStateInner {
...
+        let mut payref_status_by_tx_id = std::collections::HashMap::new();
+        
+        // Get current tip height for confirmation calculations
+        let current_tip_height = match self.wallet.base_node_service.get_chain_metadata().await {
+            Ok(Some(metadata)) => metadata.best_block_height(),
+            _ => 0, // Default to 0 if we can't get chain height
+        };
+        
+        // Process all outputs (unspent and spent) to build the lookup map
+        for output in unspent_outputs.into_iter().chain(spent_outputs.into_iter()) {
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/src/ui/state/app_state.rs]: calculate it directly from the tx.

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 14]
```diff
@@ -748,7 +889,36 @@ impl LMDBDatabase {
...
+                Ok(()) => {
+                    debug!(target: LOG_TARGET, "Successfully deleted PayRef for output {}", output_hash.to_hex());
+                },
+                Err(ChainStorageError::ValueNotFound { .. }) => {
+                    // PayRef not found - this is acceptable as it might not have been created in older versions
+                    debug!(target: LOG_TARGET, "PayRef not found for output {} - likely from older version", output_hash.to_hex());
+                },
+                Err(e) => {
+                    // Actual database error - log warning but don't fail the transaction
+                    warn!(target: LOG_TARGET, "Failed to delete PayRef for output {}: {}", output_hash.to_hex(), e);
```
Comment (SWvheerden) [inline base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs]: on this we need to error, and let the node retry the tx , or what ever higher end logic there is. 
if we allow this, we cause the db to be in a broken state

File: base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs [hunk 15]
```diff
@@ -1038,6 +1208,29 @@ impl LMDBDatabase {
...
+                    txn,
+                    &self.payref_to_output_index,
+                    &payref_bytes,
+                    "payref_to_output_index",
+                ) {
+                    Ok(()) => debug!(target: LOG_TARGET, "Deleted PayRef during reorg for output {}", output_hash.to_hex()),
+                    Err(ChainStorageError::ValueNotFound { .. }) => {
+                        // Expected case for outputs that didn't have PayRefs
+                    },
+                    Err(e) => warn!(target: LOG_TARGET, "Failed to delete PayRef during reorg for output {}: {}", output_hash.to_hex(), e),
```
Comment (SWvheerden) [inline base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs]: should error

File: base_layer/wallet/src/output_manager_service/storage/models.rs [hunk 16]
```diff
@@ -87,6 +89,104 @@ impl DbWalletOutput {
...
+                    // transactions. This is because:
+                    // 1. Genuine received payments usually use OneSided or StealthOneSided sources
+                    // 2. Change outputs are created with Standard source during transaction building
+                    //
+                    // TODO: For better accuracy, we could check if received_in_tx_id corresponds to
+                    // an outbound transaction we initiated by querying the outbound_transactions table.
+                    PaymentDirection::SentChange
+                } else {
+                    PaymentDirection::Sent
+                }
```
Comment (SWvheerden) [inline base_layer/wallet/src/output_manager_service/storage/models.rs]:
```suggestion
if self.payment_id == PaymetID::TransactionInfo{
PaymentDirection::SentChange}else {PaymentDirection::Sent}
```
This is not maybe, this is guaranteed to be accurate.

File: base_layer/wallet/src/transaction_service/handle.rs [hunk 17]
```diff
@@ -208,6 +208,16 @@ pub enum TransactionServiceRequest {
     GetFeePerGramStatsPerBlock {
         count: usize,
     },
+    /// Get all PayRefs for a specific transaction
+    GetTransactionPayRefs(TxId),
+    /// Get transaction details for a PayRef (enhanced with multiple recipients)
+    GetPaymentByReference([u8; 32]),
```
Comment (SWvheerden) [inline base_layer/wallet/src/transaction_service/handle.rs]:
```suggestion
    GetPaymentByReference(FixedHash),
```

File: base_layer/wallet/src/transaction_service/protocols/transaction_send_protocol.rs [hunk 18]
```diff
@@ -986,6 +993,50 @@ where
...
+                    match self.resources.output_manager_service.is_change_output(output).await {
+                        Ok(true) => change_hashes.push(output_hash),
+                        Ok(false) => {
+                            // It's ours but not change - could be a self-send, treat as sent
+                            sent_hashes.push(output_hash);
+                        },
+                        Err(e) => {
+                            warn!(target: LOG_TARGET, "Error checking if output is change: {:?}", e);
+                            // Default to treating as sent to be safe
+                            sent_hashes.push(output_hash);
```
Comment (SWvheerden) [inline base_layer/wallet/src/transaction_service/protocols/transaction_send_protocol.rs]: I think we should error here, something is wrong, don't save wrong results into the database

File: base_layer/wallet/src/transaction_service/service.rs [hunk 19]
```diff
@@ -3911,6 +3923,185 @@ where
         }
         Ok(())
     }
+
+    /// Get PayRefs for a specific transaction with verification
+    fn get_transaction_payrefs(&self, tx_id: TxId) -> Result<Vec<[u8; 32]>, TransactionServiceError> {
```
Comment (SWvheerden) [inline base_layer/wallet/src/transaction_service/service.rs]:
```suggestion
    fn get_transaction_payrefs(&self, tx_id: TxId) -> Result<Vec<Fixedhash>, TransactionServiceError> {
```

File: base_layer/wallet/src/transaction_service/storage/sqlite_db.rs [hunk 20]
```diff
@@ -68,6 +68,20 @@ use crate::{
...
+
+/// Deserialize a vector of HashOutput from database storage
+fn deserialize_output_hashes(data: Option<Vec<u8>>) -> Result<Vec<HashOutput>, TransactionStorageError> {
+    match data {
+        Some(bytes) => bincode::deserialize(&bytes)
+            .map_err(|e| TransactionStorageError::BincodeDeserialize(e.to_string())),
+        None => Ok(Vec::new()),
+    }
+}
+
```
Comment (SWvheerden) [inline base_layer/wallet/src/transaction_service/storage/sqlite_db.rs]: We dont need to use bincode here, its already in raw byte form

File: base_layer/wallet/src/transaction_service/storage/sqlite_db.rs [hunk 21]
```diff
@@ -1409,6 +1426,7 @@ impl InboundTransactionSql {
             last_send_timestamp: i.last_send_timestamp.map(|t| t.naive_utc()),
             payment_id: Some(i.payment_id.to_bytes()),
             user_payment_id,
+            received_output_hashes: Some(serialize_output_hashes(&i.received_output_hashes)?),
```
Comment (SWvheerden) [inline base_layer/wallet/src/transaction_service/storage/sqlite_db.rs]:
```suggestion
            received_output_hashes: Some(i.received_output_hashes.to_vec()),
```

File: base_layer/wallet_ffi/src/lib.rs [hunk 22]
```diff
@@ -231,6 +236,49 @@ pub type TariComAndPubSignature = ComAndPubSignature;
 pub type TariUnblindedOutput = UnblindedOutput;
 pub struct TariUnblindedOutputs(Vec<UnblindedOutput>);
 
+/// PayRef Configuration FFI Types
```
Comment (SWvheerden) [inline base_layer/wallet_ffi/src/lib.rs]: Can we move all of this to a separate file we did with `ffi_basenode_state`

Comment (SWvheerden) [reply base_layer/wallet_ffi/src/lib.rs]: We can also add from method to convert from the tari core methods to these c structs

**Review by SWvheerden — CHANGES_REQUESTED** [review]
Looking good, 
few comments I think we need to address
and I think rather than use `[u8;32]` everywhere for the payref, we should simply replace that with `Fixedhash`  as that already wraps that and then we get all the helper methods and type safety

File: base_layer/wallet_ffi/src/lib.rs [hunk 23]
```diff
@@ -9469,136 +9521,237 @@ pub unsafe extern "C" fn wallet_cancel_pending_transaction(
...
+            // Output hash not found
+            ptr::null_mut()
+        },
         Err(e) => {
             *error_out = LibWalletError::from(WalletError::TransactionServiceError(e)).code;
-            false
+            ptr::null_mut()
         },
     }
 }
```
[the remark this answers is not in the document]
Comment (fluffypony) [reply base_layer/wallet_ffi/src/lib.rs]: The FFI doesn't have a database...?

File: applications/minotari_app_grpc/proto/base_node.proto [hunk 24]
```diff
@@ -559,3 +561,48 @@ message LivenessResult{
...
+    uint64 spent_height = 7;
+    // Block hash where output was spent (if spent)
+    bytes spent_block_hash = 8;
+    // Transaction output amount (may be hidden for confidential transactions)
+    optional uint64 revealed_amount = 9;
+}
+
+
+// Status of payment reference lookup
+enum PaymentReferenceStatus {
```
Comment (stringhandler) [inline applications/minotari_app_grpc/proto/base_node.proto]: This doesn't seem to be used
